x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/assignment/TrainedModelAssignmentClusterService.java

```diff
@@ -555,6 +556,148 @@ private boolean isScalingPossible(List<DiscoveryNode> nodes) {
+                    }
+                }
+            }),
+            listener::onFailure
+        );
+
+        adjustNumberOfAllocations(clusterState, existingAssignment, numberOfAllocations, updatedStateListener);
+    }
+
+    private void adjustNumberOfAllocations(
+        ClusterState clusterState,
+        TrainedModelAssignment assignment,
+        int numberOfAllocations,
+        ActionListener<ClusterState> listener
+    ) {
+        threadPool.executor(MachineLearning.UTILITY_THREAD_POOL_NAME).execute(() -> {
+            if (numberOfAllocations > assignment.getTaskParams().getNumberOfAllocations()) {
+                increaseNumberOfAllocations(clusterState, assignment, numberOfAllocations, listener);
+            } else {
+                decreaseNumberOfAllocations(clusterState, assignment, numberOfAllocations, listener);
+            }
+        });
+    }
+
+    private void increaseNumberOfAllocations(
+        ClusterState clusterState,
+        TrainedModelAssignment assignment,
+        int numberOfAllocations,
+        ActionListener<ClusterState> listener
+    ) {
+        try {
+            final ClusterState updatedClusterState = update(
+                clusterState,
+                TrainedModelAssignmentMetadata.builder(clusterState)
+                    .updateAssignment(
+                        assignment.getModelId(),
+                        TrainedModelAssignment.Builder.fromAssignment(assignment).setNumberOfAllocations(numberOfAllocations)
+                    )
+            );
+            TrainedModelAssignmentMetadata.Builder rebalancedMetadata = rebalanceAssignments(updatedClusterState, Optional.empty());
+            if (isScalingPossible(getAssignableNodes(clusterState)) == false
+                && rebalancedMetadata.getAssignment(assignment.getModelId()).build().totalTargetAllocations() < numberOfAllocations) {
+                listener.onFailure(
+                    new ElasticsearchStatusException(
+                        "Could not update deployment because there are not enough resources to provide all requested allocations",
+                        RestStatus.TOO_MANY_REQUESTS
+                    )
+                );
+            } else {
+                listener.onResponse(update(clusterState, rebalancedMetadata));
+            }
+        } catch (Exception e) {
+            listener.onFailure(e);
+        }
+    }
+
+    private void decreaseNumberOfAllocations(
+        ClusterState clusterState,
+        TrainedModelAssignment assignment,
+        int numberOfAllocations,
+        ActionListener<ClusterState> listener
+    ) {
+        TrainedModelAssignment.Builder updatedAssignment = numberOfAllocations < assignment.totalTargetAllocations()
+            ? new AllocationReducer(assignment, nodeAvailabilityZoneMapper.getMlNodesByAvailabilityZone()).reduceTo(numberOfAllocations)
```

Comment:
There's a potential discrepancy in cluster states used for different parts of the code here.  `nodeAvailabilityZoneMapper` keeps itself up to date by listening for cluster state updates.  But this method runs as part of a cluster state update that is supposed to work off of an exact version of the cluster state.  Maybe `NodeAvailabilityZoneMapper` should have a method that can calculate and return ML nodes by availability zone from a specific `ClusterState` that's passed to it?  Or maybe there's a better way to solve this problem.

Reply:
I have raised https://github.com/elastic/elasticsearch/pull/90770 to address this.

Reply:
I have now rebased this PR and am using the new method where we pass the specific cluster state